src/Features/Core/Portable/EditAndContinue/EditAndContinueDiagnosticUpdateSource.cs

```diff
@@ -0,0 +1,119 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.EditAndContinue
+{
+    [Export(typeof(EditAndContinueDiagnosticUpdateSource))]
+    [Shared]
+    internal sealed class EditAndContinueDiagnosticUpdateSource : IDiagnosticUpdateSource
+    {
+        [ImportingConstructor]
+        public EditAndContinueDiagnosticUpdateSource(IDiagnosticUpdateSourceRegistrationService registrationService)
+        {
+            registrationService.Register(this);
+        }
+
+        // for testing
+        internal EditAndContinueDiagnosticUpdateSource()
```

Comment:
❗️ MEF-exported parts should not be directly instantiated for testing (it violates the single-instance guarantee). This constructor should be removed. The previous constructor will be used when the part is obtained from the ExportProvider for use in testing.

Reply:
I am not importing this via MEF in unit tests.

---
In reply to: [270628092](https://github.com/dotnet/roslyn/pull/34027#discussion_r270628092) [](ancestors = 270628092)

Reply:
➡️ This will be flagged and corrected by the new analyzers in #35439

Reply:
I don't think we should be forcing tests to create types like this via MEF. If I want to write a simple unit test for a method in this type, I don't want the overhead of setting up MEF composition.
Also, some types do not assume that they must be created via MEF. They work perfectly find if they are not.

---
In reply to: [287487306](https://github.com/dotnet/roslyn/pull/34027#discussion_r287487306) [](ancestors = 287487306)